assignments/burmanm/week-2.scm

```diff
@@ -0,0 +1,394 @@
+;; Procedures and the Processes They Generate (1.2)
+
+;; Exercise 1.9
+;; First one is recursive as we need to keep the stack since we're unable to calculate the
+;; value without knowing the subprocesses return value.
+;;
+;; (+ 4 5)
+;;   (inc (+ 3 5))
+;;     (inc (inc (+ 2 5)))
+;;       (inc (inc (inc (+ 1 5))))
+;;         (inc (inc (inc (inc (+ 0 5)))))
+;;           (inc (inc (inc (inc  5))))
+;;         (inc (inc (inc 6)))
+;;       (inc (inc 7))
+;;     (inc 8)
+;;   9
+;;
+;; Second process is iterative, the state can be maintained by only 'a' and 'b', without the need
+;; to know any previous recursive call states.
+;;
+;; The latter example with (+ 4 5) generates the calls:
+;; (+ 4 5)
+;; (+ 3 6)
+;; (+ 2 7)
+;; (+ 1 8)
+;; (+ 0 9)
+;; b -> (9)
+
+;; Exercise 1.10
+
+;; 1024
+;; 65536
+;; 65536
+;;
+;; Omitted mathematical stuff
+;;
+;; Exercise 1.11
+;;
+;;
+(define (f-recursive n)
+  (if (< n 3)
+      n
+      (+ (f (- n 1)) (* 2 (f (- n 2))) (* 3 (f (- n 3))))))
+
+(define (f-iterative n)
+  (define (iter count psum ppsum pppsum)
+    (if (> count n)
+	psum
+	(iter (+ count 1) (+ psum (* 2 ppsum) (* 3 pppsum)) psum ppsum)))
+  (iter 3 2 1 0))
+
+;; Exercise 1.12
+;;
+;; Calculates single position only, not the whole line / tree / etc
+(define (pascal n i)
+  (cond ((or (< i 1) (> i n)) -1)
+	((= i 1) 1)
+	((= i n) 1)
+	(else (+ (pascal (- n 1) (- i 1)) (pascal (- n 1) i)))))
+
+;;
+;; Exercise 1.13
+;;
+;; Skipping this one, it's just mathematical induction.
+;;
+
+;; Exercise 1.14
+;; No time, it increases with some exponential function.
+;;
+;; Exercise 1.15
+;;
+;; 5 times and it's O(log n) as we can see the sine value approaching with ~sqrt(n) changes.
+
+;; Exercise 1.16
+(define (fast-expt-iterative b n)
+  (define (calc-iter a x z)
+    (cond ((= z 0) a)
+	  ((even? z) (calc-iter a (square x) (/ z 2)))
+	  (else (calc-iter (* a x) x (- z 1)))))
+  (calc-iter 1 b n))
+
+;; Exercise 1.17
+(define (double a) (+ a a))
+(define (halve a) (/ a 2))
+(define (fast-mult a b)
+  (cond ((= b 1) a)
```

Comment:
It seems like it's missing a test for b == 0, otherwise it fails to evaluate: (fast-mult 4 0)